### PR TITLE
[2.x] Use snappy instead of `ParallelGzipOutputStream`

### PR DESCRIPTION
**Pull request by Friendseeker**
### Issue

`ParallelGzipOutputStream` occasionally enters deadlock. c.c. https://github.com/sbt/zinc/pull/1419#issuecomment-2394831504

### Fix

Use [Snappy](https://github.com/xerial/snappy-java) instead of `ParallelGzipOutputStream`.

### Performance Impact

| Format                                         | Write Time (ms/op) | Read Time (ms/op) |
|------------------------------------------------|--------------------|-------------------|
| ConsistentBinaryWithParallelGZIP                               | 284.774 ± 6.749    | 160.292 ± 1.284   |
| ConsistentBinaryWithParallelGZIP (No Sort)                     | 202.954 ± 1.903    | N/A               |
| ConsistentBinaryWithSnappy                     | 267.960 ± 7.312    | 152.017 ± 1.065   |
| ConsistentBinaryWithSnappy (No Sort)           | 219.906 ± 4.312    | N/A               |
| ConsistentBinaryWithStandardGZIP               | 581.248 ± 3.852    | 158.442 ± 1.322   |
| ConsistentBinaryWithStandardGZIP (No Sort)     | 625.460 ± 5.160    | N/A               |

c.c. https://github.com/sbt/zinc/pull/1419#issuecomment-2394886677

Benchmark suggests `Snappy` is slightly faster than `ParallelGzipOutputStream`.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -261,7 +261,10 @@ lazy val zincPersist = (projectMatrix in internalPath / "zinc-persist")
   .dependsOn(zincCore, zincCompileCore, zincCore % "test->test")
   .settings(
     name := "zinc Persist",
-    libraryDependencies += sbinary,
+    libraryDependencies ++= Seq(
+      sbinary,
+      snappy,
```
Comment (eed3si9n) [inline build.sbt:266]: Given the wide use of Zinc, I would rather avoid introducing compression library maintained by a single person (e.g. https://en.wikipedia.org/wiki/XZ_Utils_backdoor). The author is a long-time community member and this isn't meant to be a FUD, but generally I think we should be careful about introducing [native code](https://github.com/xerial/snappy-java/blob/master/src/main/resources/org/xerial/snappy/native/Mac/aarch64/libsnappyjava.dylib) that we didn't write unless it's heavily audited.

Comment (Friendseeker) [reply build.sbt:266]: Makes sense. I shall do more research on what other compression libraries are available. Hopefully there's one that we can better trust.

**Review by eed3si9n — CHANGES_REQUESTED** [review]
I'll mark this as request changes.

**Comment by Friendseeker** [comment]
@eed3si9n I did some research and I failed to find any fast compression libraries satisfying the criteria. There was some seeming promising candidates such as https://github.com/apache/commons-compress but they actually use `snappy-java` or `lz4-java` underneath.

https://github.com/airlift/aircompressor is maintained by multiple devs actively and is done in pure java, but it requires minimum JVM 22.

https://github.com/flanglet/kanzi is done by pure Java, supports Java 8 but only has a single dev and has no release on maven central.

`lz4-java` actually has a pure Java implementation that supports Java 8 (https://search.maven.org/search?q=g:org.lz4%20a:lz4-pure-java). I uncompressed the release jar from maven central and can confirm there's no native code hidden in it. The issue is that `lz4-java` currently has no active maintainers and the last release was made 3 years ago.

As for fixing the underlying issue in `ParallelGzipOutputStream`, @Ichoran was looking into implementation of `ParallelGzipOutputStream` but found no issue. Ichoran proposed rewriting `ParallelGzipOutputStream` using just threads. But even if Ichoran devoted effort to rewritten it & quash the issue, I am still concerned about maintainability of `ParallelGzipOutputStream`, as concurrent codes are orders of magnitude harder to maintain & debug. If `ParallelGzipOutputStream` ever has issue again, then... we cannot exactly keep rewriting it from ground up again.

I do feel like at this point, there's no perfect solution to the problem. We have to make tradeoff and pick the least worst solution from the following

- Use `java.util.zip.GZIPOutputStream`
  - Secure, but very slow
- Use snappy-java
  - Contains unaudited native code, but fast
- Use lz4-pure-java
  - Does not contain native code, fast, but under no active maintenance and is outdated
- Ask Ichoran or other volunteers to fix `ParallelGzipOutputStream` 
  - May or may not identify the root cause of the issue, and we have to bear the burden of maintaining `ParallelGzipOutputStream`

**Comment by Friendseeker** [comment]
Some recent comments on the issue from discord from Haoyi & Ichoran. Listing them here for documentation purpose.

https://discord.com/channels/632150470000902164/632150470000902166/1293090991749992480

https://discord.com/channels/632150470000902164/632150470000902166/1293096900073095168

**Comment by Friendseeker** [comment]
As extra info, it seems like `snappy-java`'s code is at least audited under some extent.

For example in 2023 a CVE on `snappy-java` (https://github.com/advisories/GHSA-qcwq-55hx-v3vh) was discovered by security researcher Shachar Menashe from JFrog.

**Comment by eed3si9n** [comment]
Also in the big picture, we should consider the pros and cons of introducing native code vs perf gain in the wallclock time. We're talking about 200ms here where many of the compile takes significantly longer than that, right?

One strategy we might try could be start parallel one and sequential one at the same time with different file name, and if sequential one is faster, cancel the parallel one and mv the file.

**Comment by Friendseeker** [comment]
> Also in the big picture, we should consider the pros and cons of introducing native code vs perf gain in the wallclock time. We're talking about 200ms here where many of the compile takes significantly longer than that, right?

I have never engaged in development of very large Scala monorepo. @szeiger Would you like to join the discussion? Basically there's some unknown concurrency issue caused by usage of `ParallelGzipOutputStream` and we are discussing how to address the problem, options we can take and pro & con of each approaches. 

Btw citing Stefan's [comment](https://discord.com/channels/632150470000902164/1035522798027800616/1181663007835705395), analysis serialization with old format takes 7s in their internal repo.

> One strategy we might try could be start parallel one and sequential one at the same time with different file name, and if sequential one is faster, cancel the parallel one and mv the file.

While this is a valid strategy, I do feel like this is a really hacky solution that does not address the underlying concurrency issue. If we do need a temporary solution to the problem, I personally think that just using `java.util.zip.GZIPOutputStream` or `java.util.zip.DeflaterOutputStream` might be a acceptable interim solution.

Just as Eugene said, even for large projects like Scala Compiler, we are only talking about 300ms of time saving. The optimization only matters for a small amount of very large (likely proprietary) Scala projects, for which any small code change would take at least 7s serialization time. In this case serialization overhead may be significantly longer than compile time.

**Comment by Ichoran** [comment]
I should be able to finish the raw-thread version of Stefan's algorithm within the next 2-3 hours, or, failing that, tomorrow.

**Comment by Ichoran** [comment]
Yup, got it!  [Gist here](https://gist.github.com/Ichoran/a4c797f16a968e9e85c4492efc6cb474); I don't have time to make a PR right now.

**Comment by Friendseeker** [comment]
> Yup, got it! [Gist here](https://gist.github.com/Ichoran/a4c797f16a968e9e85c4492efc6cb474); I don't have time to make a PR right now.

Stress test CI triggered! https://github.com/Friendseeker/zinc/actions/runs/11250109731

---

Yay it passed!

**Comment by Friendseeker** [comment]
Here's the Zinc benchmark result. `ConsistentBinaryWithParallelGZIP` refers to Stefan's implementation, while `ConsistentBinaryWithNewGZIP` refers to Ichoran's implementation. `ConsistentBinaryWithStandardGZIP` refers to Java's built in gzip output stream.

<table><thead><tr><th>Format</th><th>Write Time (ms/op)</th><th>Read Time (ms/op)</th></tr></thead><tbody><tr><td>ConsistentBinaryWithParallelGZIP</td><td>287.663 ± 17.311</td><td>169.598 ± 3.594</td></tr><tr><td>ConsistentBinaryWithParallelGZIP (No Sort)</td><td>210.229 ± 7.558</td><td>N/A</td></tr><tr><td>ConsistentBinaryWithSnappy</td><td>228.724 ± 8.195</td><td>134.769 ± 2.093</td></tr><tr><td>ConsistentBinaryWithSnappy (No Sort)</td><td>160.253 ± 13.871</td><td>N/A</td></tr><tr><td>ConsistentBinaryWithStandardGZIP</td><td>598.675 ± 24.225</td><td>N/A</td></tr><tr><td>ConsistentBinaryWithStandardGZIP (No Sort)</td><td>631.005 ± 8.725</td><td>N/A</td></tr><tr><td>ConsistentBinaryWithNewGZIP</td><td>272.501 ± 2.371</td><td>N/A</td></tr><tr><td>ConsistentBinaryWithNewGZIP (No Sort)</td><td>225.453 ± 9.408</td><td>N/A</td></tr></tbody></table>

Here's the size benchmark. ConsistentBinary (Parallel GZIP) refers to Stefan/Ichoran's implementation (since both have binary identical output). ConsistentBinary (Stock GZIP) refers to Java's built in gzip output stream.

<table><thead><tr><th>Format</th><th>Size</th></tr></thead><tbody><tr><td>ConsistentBinary (Parallel GZIP)</td><td>3282k</td></tr><tr><td>ConsistentBinary (GZIP nosort)</td><td>4016k</td></tr><tr><td>ConsistentBinary (snappy)</td><td>4315k</td></tr><tr><td>ConsistentBinary (snappy nosort)</td><td>5314k</td></tr><tr><td>ConsistentBinary (Stock GZIP) total</td><td>3180k</td></tr><tr><td>ConsistentBinary (Stock GZIP nosort)</td><td>3865k</td></tr></tbody></table>

c.c. https://github.com/Friendseeker/zinc/actions/runs/11269808474/job/31339169008

In summary Ichoran's implementation is a drop in replacement of Stefan's implementation. I shall make another PR incorporating Ichoran's implementation to Zinc so Ichoran don't have to spend extra time making a PR.

**Comment by Ichoran** [comment]
@Friendseeker - Mention me when you get the PR together.  Thanks!
